client/receiver.go

```diff
@@ -0,0 +1,288 @@
+			switch contentType {
+			case constants.HeaderValueJson:
+				body, err = json.MarshalIndent(reqBody, "", "  ")
+				if err != nil {
+					return err
+				}
+				break
+			case constants.HeaderValueXml:
+				body, err = xml.Marshal(reqBody)
+				if err != nil {
+					return err
+				}
+
+				break
+			default:
+				log.Printf("Content-Type is not application/json nor application/xml\n")
+			}
+		}
+
+		log.Printf("Request body:\n%s\n", string(body))
+		if needMarshal {
+			req.Body(body)
+		} else {
+			req.Body(reqBody)
+		}
+	}
+
+	//init header
+	if headers != nil {
+		for k, v := range headers {
+			req.Header(k, v)
+		}
+	}
+
+	// Get http response.
+	resp, err := req.Response()
+	if err != nil {
+		return err
+	}
+
+	if 400 <= resp.StatusCode && resp.StatusCode <= 599 {
+		log.Printf("Response statusCode: %+v\n", resp.StatusCode)
+		return NewHTTPError(resp.StatusCode, "")
+	}
+
+	rbody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		log.Printf("ioutil.ReadAll failed, err: %+v\n", err)
+		return err
+	}
+
+	defer resp.Body.Close()
+	log.Printf("Response Status: %s\nResponse Body:\n%s\n", resp.Status, string(rbody))
+	if (nil == rbody) || ("" == string(rbody)) {
+		return nil
+	}
+
+	if "" == outFileName {
+		if nil == respBody {
+			return nil
+		}
+
+		var respContentType string
+		respContentTypes, ok := resp.Header["Content-Type"]
+		log.Printf("ok=%+v, respContentTypes=%+v, len=%v\n", ok, respContentTypes, len(respContentTypes))
+
+		if ok && len(respContentTypes) > 0 {
+			respContentType = respContentTypes[0]
+		}
+
+		switch respContentType {
+		case constants.HeaderValueJson:
+			if err = json.Unmarshal(rbody, respBody); err != nil {
+				return fmt.Errorf("failed to unmarshal result message: %v", err)
+			}
+			log.Printf("application/json, respBody=%+v\n", respBody)
+			break
+		case constants.HeaderValueXml, "text/xml; charset=utf-8":
+			if err = xml.Unmarshal(rbody, respBody); err != nil {
+				return fmt.Errorf("failed to unmarshal result message: %v", err)
+			}
+			log.Printf("application/xml, respBody=%+v\n", respBody)
+			break
+		default:
+			log.Printf("Failure to process the response body!")
+		}
+	} else {
+		path := fmt.Sprintf("./%s", outFileName)
+		file, err := os.Create(path)
+		if err != nil {
+			log.Printf("Failed to create file:%+v\n", err)
+		}
+		defer file.Close()
+
+		n, err := file.Write(rbody)
+		if err != nil {
+			log.Printf("Failed to Write file,err:%+v\n, n:%+v\n", err, n)
+		}
+		log.Printf("Save file successfully, n:%+v\n", n)
+	}
+
+	return nil
+}
+
+type receiver struct{}
+
+func (*receiver) Recv(url string, method string, headers HeaderOption,
+	reqBody interface{}, respBody interface{}, needMarshal bool, outFileName string) error {
+	return request(url, method, headers, reqBody, respBody, needMarshal, outFileName)
+}
+
+// NewKeystoneReciver implementation
+func NewKeystoneReciver(auth *KeystoneAuthOptions) Receiver {
+	k := &KeystoneReciver{Auth: auth}
+	k.GetToken()
```

Comment:
It may get errors, when geting token, so suggestion you to handle this error,  and return this error to upper function.